### PR TITLE
Add infrastructure for too many Includes warning.

### PR DESCRIPTION
**Pull request by ajcvickers**
Part of #19933

### CODE REVIEW

**Comment by ajcvickers** [comment]
@smitpatel My part of #19933. Warning is not wired up yet--need you for that.

**Comment by smitpatel** [comment]
What is our cut-off to generate warning?

**Comment by ajcvickers** [comment]
@smitpatel I'll just let you take over this PR in whatever way you want, since I can't get approval to merge until the complete change is done anyway.

**Comment by AndriySvyryd** [comment]
@smitpatel If the query has two or more collection includes or collection nav expansions then we should generate the warning

File: src/EFCore.Relational/Diagnostics/RelationalLoggerExtensions.cs [hunk 1]
```diff
@@ -3701,6 +3702,47 @@ private static string ModelValidationKeyDefaultValueWarning(EventDefinitionBase
                 p.Property.DeclaringEntityType.DisplayName());
         }
 
+        /// <summary>
+        ///     Logs for the <see cref="RelationalEventId.TooManyIncludesWarning" /> event.
+        /// </summary>
+        /// <param name="diagnostics"> The diagnostics logger to use. </param>
+        /// <param name="expressionPrinter"> Used to create a human-readable representation of the expression tree. </param>
+        /// <param name="queryExpression"> The query expression tree. </param>
+        public static void TooManyIncludesWarning(
```
Comment (smitpatel) [inline src/EFCore.Relational/Diagnostics/RelationalLoggerExtensions.cs]: The place we know that there too many includes in relational, we don't have actual expression tree.

Comment (ajcvickers) [reply src/EFCore.Relational/Diagnostics/RelationalLoggerExtensions.cs]: @smitpatel I can update it to take whatever you have available.

Comment (smitpatel) [reply src/EFCore.Relational/Diagnostics/RelationalLoggerExtensions.cs]: Let's discuss in triage today.

**Comment by smitpatel** [comment]
I looked at this further, the place where we accurately know that there is multiple collection projection, we only have SelectExpression. Further, in order to print SelectExpression in most readable format, we need sql generator which we don't have. Any other place to identify this requires a public surface on places where we don't want to add public surface. (even if we add in patch, it would be throw away work).

At this point, warning seems pretty low value as if query (with or without multiple collection include), either it works just fine or perf is pretty bad and customers would know.

**Comment by ajcvickers** [comment]
@smitpatel Can we get any type of string from the SelectExpression? If so, can you post examples of what it looks like?

**Comment by smitpatel** [comment]
Can strip out shaper if needed
```
(ShapedQueryExpression: 
    QueryExpression: 
        (Projection Mapping:
        SELECT ((c.CustomerID)), ((c.Address)), ((c.City)), ((c.CompanyName)), ((c.ContactName)), ((c.ContactTitle)), ((c.Country)), ((c.Fax)), ((c.Phone)), ((c.PostalCode)), ((c.Region)), ((t.OrderID)), ((t.CustomerID)), ((t.EmployeeID)), ((t.OrderDate)), ((t.OrderID0)), ((t.ProductID)), ((t.Discount)), ((t.Quantity)), ((t.UnitPrice))
        FROM (Customers AS c)
        (LEFT JOIN (Projection Mapping:
        (
            SELECT ((o.OrderID)), ((o.CustomerID)), ((o.EmployeeID)), ((o.OrderDate)), ((o0.OrderID) AS OrderID0), ((o0.ProductID)), ((o0.Discount)), ((o0.Quantity)), ((o0.UnitPrice))
            FROM (Orders AS o)
            (LEFT JOIN (Order Details AS o0) ON ((o.OrderID) == (o0.OrderID)))
        ) AS t) ON ((c.CustomerID) == (t.CustomerID)))
        ORDER BY ((c.CustomerID) ASC), ((t.OrderID) ASC), ((t.OrderID0) ASC), ((t.ProductID) ASC))
    ShaperExpression: (IncludeExpression(
            (EntityShaperExpression: 
                EntityType: Customer
                ValueBufferExpression: 
                    (ProjectionBindingExpression: CustomerID:0,
                        Address:1,
                        City:2,
                        CompanyName:3,
                        ContactName:4,
                        ContactTitle:5,
                        Country:6,
                        Fax:7,
                        Phone:8,
                        PostalCode:9,
                        Region:10,
                    )
                IsNullable: False
            ), 
            (RelationalCollectionShaper:
                CollectionId: 0
                ParentIdentifier:new object[]{ (ProjectionBindingExpression: 0) }
                OuterIdentifier:new object[]{ (ProjectionBindingExpression: 0) }
                SelfIdentifier:new object[]{ (object)(ProjectionBindingExpression: 11) }
                InnerShaper:(IncludeExpression(
                    (EntityShaperExpression: 
                        EntityType: Order
                        ValueBufferExpression: 
                            (ProjectionBindingExpression: OrderID:11,
                                CustomerID:12,
                                EmployeeID:13,
                                OrderDate:14,
                            )
                        IsNullable: True
                    ), 
                    (RelationalCollectionShaper:
                        CollectionId: 1
                        ParentIdentifier:new object[]{ (object)(ProjectionBindingExpression: 11) }
                        OuterIdentifier:new object[]{ (object)(ProjectionBindingExpression: 11) }
                        SelfIdentifier:new object[]
                        { 
                            (object)(ProjectionBindingExpression: 15), 
                            (object)(ProjectionBindingExpression: 16) 
                        }
                        InnerShaper:(EntityShaperExpression: 
                            EntityType: OrderDetail
                            ValueBufferExpression: 
                                (ProjectionBindingExpression: OrderID:15,
                                    ProductID:16,
                                    Discount:17,
                                    Quantity:18,
                                    UnitPrice:19,
                                )
                            IsNullable: True
                        )
                        Navigation: OrderDetails
                    ), OrderDetails)
                )
                Navigation: Orders
            ), Orders)
        )
)
```

**Comment by ajcvickers** [comment]
@smitpatel Thanks. Let's discuss again in triage. :-)

**Comment by AndriySvyryd** [comment]
This still has value. If one needs to identify the query that triggers the warning they can just set it to throw.

**Comment by smitpatel** [comment]
> This still has value. 

You mean warning or exception message? Because in today's triage while discussing, people said, if we cannot give good indication of which query then it is not useful.

**Comment by AndriySvyryd** [comment]
> You mean warning or exception message? 

It's one and the same, depends on user configuration.

> Because in today's triage while discussing, people said, if we cannot give good indication of which query then it is not useful.

If it throws then it's clear

File: src/EFCore.Relational/Properties/RelationalStrings.resx [hunk 2]
```diff
@@ -250,6 +250,10 @@
     <value>Possible unintended use of a potentially throwing aggregate method (Min, Max, Average) in a subquery. Client evaluation will be used and operator will throw if no data exists. Changing the subquery result type to a nullable type will allow full translation.</value>
     <comment>Warning RelationalEventId.QueryPossibleExceptionWithAggregateOperatorWarning</comment>
   </data>
+  <data name="LogTooManyIncludesWarning" xml:space="preserve">
+    <value>The query '{query}' uses 'Include' for more than one collection navigation property. Including multiple collections can result in slow query performance. See ... for more information.</value>
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Properties/RelationalStrings.resx]: Don't forget to replace the `...`
Add "To identify the query that's triggering this warning call `.ConfigureWarnings(w => w.Throw(RelationalEventId.LogTooManyIncludesWarning))`"

**Comment by ajcvickers** [comment]
@smitpatel What I intended to say in triage was, "if we cannot give _any_ indication of which query then it is not useful." :trollface:

**Comment by ajcvickers** [comment]
Blocked for now until we have the new code in place.

**Comment by ajcvickers** [comment]
@smitpatel Is this PR now obsolete?

**Comment by AndriySvyryd** [comment]
How else will the users find out that they need to call the new API?

**Comment by ajcvickers** [comment]
@AndriySvyryd I meant specifically this PR, not the concept of having a warning.

**Comment by smitpatel** [comment]
So should I put this thing to master and generate warning when not split query and more than one collection is being projected/included?

**Comment by ajcvickers** [comment]
@smitpatel Yes. :-D

**Comment by smitpatel** [comment]
Superseded by #21419

**Pull request closed without merging**